pkg/objstore/swift/swift.go

```diff
@@ -82,215 +201,135 @@ func (c *Container) Name() string {
+func (c *Container) get(name string, headers swift.Headers, checkHash bool) (io.ReadCloser, error) {
 	if name == "" {
-		return nil, errors.New("error, empty container name passed")
+		return nil, errors.New("object name cannot be empty")
+	}
+	file, _, err := c.connection.ObjectOpen(c.name, name, checkHash, headers)
+	if err != nil {
+		return nil, errors.Wrap(err, "swift open object")
 	}
-	response := objects.Download(c.client, c.name, name, nil)
-	return response.Body, response.Err
+	return file, err
 }
 
-// GetRange returns a new range reader for the given object name and range.
-func (c *Container) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
-	lowerLimit := ""
-	upperLimit := ""
-	if off >= 0 {
-		lowerLimit = fmt.Sprintf("%d", off)
-	}
-	if length > 0 {
-		upperLimit = fmt.Sprintf("%d", off+length-1)
-	}
-	options := objects.DownloadOpts{
-		Newest: true,
-		Range:  fmt.Sprintf("bytes=%s-%s", lowerLimit, upperLimit),
+// Get returns a reader for the given object name.
+func (c *Container) Get(_ context.Context, name string) (io.ReadCloser, error) {
+	return c.get(name, swift.Headers{}, true)
+}
+
+func (c *Container) GetRange(_ context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	// Set Range HTTP header, see the docs https://docs.openstack.org/api-ref/object-store/?expanded=show-container-details-and-list-objects-detail,get-object-content-and-metadata-detail#id76.
+	bytesRange := fmt.Sprintf("bytes=%d-", off)
+	if length != -1 {
+		bytesRange = fmt.Sprintf("%s%d", bytesRange, off+length-1)
 	}
-	response := objects.Download(c.client, c.name, name, options)
-	return response.Body, response.Err
+	return c.get(name, swift.Headers{"Range": bytesRange}, false)
 }
 
 // Attributes returns information about the specified object.
-func (c *Container) Attributes(ctx context.Context, name string) (objstore.ObjectAttributes, error) {
-	response := objects.Get(c.client, c.name, name, nil)
-	headers, err := response.Extract()
+func (c *Container) Attributes(_ context.Context, name string) (objstore.ObjectAttributes, error) {
+	if name == "" {
+		return objstore.ObjectAttributes{}, errors.New("object name cannot be empty")
+	}
+	info, _, err := c.connection.Object(c.name, name)
 	if err != nil {
-		return objstore.ObjectAttributes{}, err
+		return objstore.ObjectAttributes{}, errors.Wrap(err, "swift get object attributes")
 	}
-
 	return objstore.ObjectAttributes{
-		Size:         headers.ContentLength,
-		LastModified: headers.LastModified,
+		Size:         info.Bytes,
+		LastModified: info.LastModified,
 	}, nil
 }
 
 // Exists checks if the given object exists.
-func (c *Container) Exists(ctx context.Context, name string) (bool, error) {
-	err := objects.Get(c.client, c.name, name, nil).Err
-	if err == nil {
-		return true, nil
-	}
-
-	if _, ok := err.(gophercloud.ErrDefault404); ok {
-		return false, nil
+func (c *Container) Exists(_ context.Context, name string) (bool, error) {
```

Comment:
None of those support context cancellation?

Reply:
Unfortunately not. The connection lets you pass own transport, but that does not help :/